improve location-awareness (#3825)
pingcap/docs: schedule-replicas-by-topology-labels.md

---
title: Schedule Replicas by Topology Labels

> category: how-to

Reviewer: We no longer use `category` in metadata. Please refer to other .md files.

> category: how-to

Author: ```suggestion
```

> aliases: ['/docs-cn/dev/how-to/deploy/geographic-redundancy/location-awareness/','/docs-cn/dev/location-awareness/']

Reviewer: @ireneontheway Why `docs-cn` aliases in docs repo?

> aliases: ['/docs-cn/dev/how-to/deploy/geographic-redundancy/location-awareness/','/docs-cn/dev/location-awareness/']

Author: ```suggestion
aliases: ['/docs/dev/how-to/deploy/geographic-redundancy/location-awareness/','/docs/dev/location-awareness/']
```

---

# Schedule Replicas by Topology Labels

> To improve the high availability and disaster tolerance of TiDB clusters, it is recommended that TiKV nodes are physically distributed as much as possible. For example, TiKV nodes can be distributed on different racks or even in different data zones. According to the topology information of TiKV, the PD scheduler automatically performs scheduling at the background to isolate the replicas of Regions as much as possible, thereby maximizing the capability for disaster recovery.

Reviewer 2: data zones -> data centers.

> To improve the high availability and disaster tolerance of TiDB clusters, it is recommended that TiKV nodes are physically distributed as much as possible. For example, TiKV nodes can be distributed on different racks or even in different data zones. According to the topology information of TiKV, the PD scheduler automatically performs scheduling at the background to isolate the replicas of Regions as much as possible, thereby maximizing the capability for disaster recovery.

Author: ```suggestion
To improve the high availability and disaster tolerance of TiDB clusters, it is recommended that TiKV nodes are physically distributed as much as possible. For example, TiKV nodes can be distributed on different racks or even in different data centers. According to the topology information of TiKV, the PD scheduler automatically performs scheduling at the background to isolate the replicas of Regions as much as possible, thereby maximizing the capability for disaster recovery.
```

## Configure `labels` based on the cluster topology

### Set the `label` configuration of TiKV

TiKV supports binding some attributes through command-line flags or configuration files in the form of key-value pairs. The attribute is called a `label`. TiKV reports its labels to PD so we can mark the TiKV location.

Assume that the topology has three structures: zone > rack > host, use these labels to set the TiKV location.

Command-line flags:

{{< copyable "" >}}

```
tikv-server --labels zone=<zone>,rack=<rack>,host=<host>
```

Configuration files:

{{< copyable "" >}}

```toml
[server]
labels = "zone=<zone>,rack=<rack>,host=<host>"
```

### Set the `location-labels` configuration of PD

According to the previous description, labels can be any key-value pairs used to describe TiKV attributes, but PD can not distinguish the location-marking labels and the structure among them. Therefore, you need set some configuration to understand the TiKV node topology.

Set the PD configuration `location-labels` through the PD configuration files.

{{< copyable "" >}}

```toml
[replication]
location-labels = ["zone", "rack", "host"]
```

> After the PD cluster is initialized, you need to use the pd-ctl tool to make online changes:

Reviewer 2: To configure `location-labels` after the PD cluster is initialized, ...
Or
If PD cluster is already initialized, you need to...

> After the PD cluster is initialized, you need to use the pd-ctl tool to make online changes:

Author: ```suggestion
If PD cluster is already initialized, you need to use the pd-ctl tool to make online changes:
```

## Configure `labels` based on the cluster topology

### Set the `labels` configuration of TiKV

TiKV supports binding some attributes through command-line flags or configuration files in the form of key-value pairs. The attribute is called a `label`. TiKV reports its labels to PD so we can mark the TiKV location.

Assume that the topology has three structures: zone > rack > host, use these labels to set the TiKV location.

Command-line flags:

{{< copyable "" >}}

```
tikv-server --labels zone=<zone>,rack=<rack>,host=<host>
```

Configuration files:

{{< copyable "" >}}

```toml
[server]
labels = "zone=<zone>,rack=<rack>,host=<host>"
```

### Set the `location-labels` configuration of PD

According to the previous description, labels can be any key-value pairs used to describe TiKV attributes, but PD can not distinguish the location-marking labels and the structure among them. Therefore, you need to set some configuration to understand the TiKV node topology.

Set the PD configuration `location-labels` through the PD configuration files.

> {{< copyable "" >}}
> 
> ```toml
> [replication]
> location-labels = ["zone", "rack", "host"]
> ```

Reviewer: ```suggestion
    {{< copyable "" >}}

    ```toml
    [replication]
    location-labels = ["zone", "rack", "host"]
    ```
```

If PD cluster is already initialized, you need to use the pd-ctl tool to make online changes:

> {{< copyable "shell-regular" >}}
> 
> ```bash
> pd-ctl config set location-labels zone,rack,host
> ```

Reviewer: ```suggestion
    {{< copyable "shell-regular" >}}

    ```bash
    pd-ctl config set location-labels zone,rack,host
    ```
```

Assume that the topology is divided into three structures through `location-labels` according to the above instructions: zone > rack > host, and the `isolation-level` is configured as follows:

{{< copyable "" >}}

```toml
[replication]
isolation-level = "zone"
```

If PD cluster is already initialized, you need to use the pd-ctl tool to make online changes:

{{< copyable "shell-regular" >}}

```bash
pd-ctl config set isolation-level zone
```

The `location-level` configuration is an array of strings, which needs to correspond to a key of `location-labels`. This parameter limits the minimum and mandatory isolation level requirements for TiKV topology clusters.

> **Note:**
>
> `isolation-level` is empty by default, which means there is no mandatory isolation level restriction. To set it, you need to configure the `location-labels` of PD, and ensure that the value of `isolation-level` must be one of `location-labels`.

### Use TiDB Ansible to deploy a cluster

You can directly set location related configuration in the `inventory.ini` file when using TiDB Ansible to deploy a cluster. `tidb-ansible` generates the corresponding TiKV and PD configuration files during deployment.

> The following example defines a double-level topology of `zone/host`. The TiKV of the cluster is distributed in three zones, with two hosts in each zone. z1 is deployed with two TiKV instances per host, and z2 and z3 are deployed with 1 instance per host.

Reviewer 2: LGTM

```
[tikv_servers]
# z1
tikv-1 labels="zone=z1,host=h1"
tikv-2 labels="zone=z1,host=h1"
tikv-3 labels="zone=z1,host=h2"
tikv-4 labels="zone=z1,host=h2"
# z2
tikv-5 labels="zone=z2,host=h1"
tikv-6 labels="zone=z2,host=h2"
# z3
tikv-7 labels="zone=z3,host=h1"
tikv-8 labels="zone=z3,host=h2"

[pd_servers:vars]
location_labels = ["zone", "host"]

> ```

Reviewer: ```suggestion
```
</details>
```

> 

Reviewer: ```suggestion
</details>

```

Assume that the number of cluster replicas is 5 (`max-replicas=5`). As there are only 3 zones in total, PD can not guarantee the isolation of each replica at this level. At this time, the PD scheduler adjusts to ensure isolation at the host level. In other words, multiple copies of a Region might be distributed in the same zone, but will not be distributed on the same host.

Under the premise of the 5-replica configuration, if z3 occurs an overall failure or isolation, and can not be restored in a period (controlled by `max-store-down-time`), PD will make up the 5 replicas through scheduling. At this time, only 3 hosts are available, so host level isolation can not be guaranteed. Therefore, multiple copies might be distributed to the same host. But if the value `isolation-level` is set to `zone` instead of empty, this specifies the minimum physical isolation requirements for Region replicas. That is to say, PD ensures that replicas of the same Region are scattered in different zones. Even if following this isolation restriction does not meet the multiple replica requirements of `max-replicas`, PD will not perform corresponding scheduling. 

> For example, there are three data centers z1, z2 and z3 in the TiKV cluster. Under the three-replica setting, PD distributes the three replicas of the same Region to these three data centers respectively. At this time, if the entire data center z1 encounters a power outage and cannot be recovered in a period, the PD will consider that the Region copy on z1 is no longer available. But because the `isolation-level` is set to `zone`, PD needs to strictly ensure that different Region replicas will not fall into the same zone. At this time, as both z2 and z3 already have replicas, PD does not perform any scheduling under the restriction of the minimum mandatory isolation level, even if there are only two replicas.

Reviewer: ```suggestion
For example, a TiKV cluster is distributed across three data zones z1/z2/z3. Each Region has three replicas as required, and PD distributes the three replicas of the same Region to these three data zones respectively. If a power outage occurs in z1 and cannot be recovered after a period of time, PD determines that the Region replicas on z1 are no longer available. However, because `isolation-level` is set to `zone`, PD needs to strictly guarantee that different replicas of the same Region will not be scheduled on the same data zone. Because both z2 and z3 already have replicas, PD will not perform any scheduling under the minimum isolation level restriction of `isolation-level`, even if there are only two replicas at this moment.
```

> Similarly, when `isolation-level` is set to `rack`, the minimum isolation level is different racks in the same data center. Under this setting, if isolation can be guaranteed at the zone level, it will be guaranteed first. Only when the zone level isolation can not be completed, will the scheduling in the same zone and the same rack be considered to avoid.

Reviewer: ```suggestion
Similarly, when `isolation-level` is set to `rack`, the minimum isolation level applies to different racks in the same data zone. With this configuration, the isolation at the zone level is guaranteed first if possible. When the isolation at the zone level cannot be guaranteed, PD tries to avoid scheduling different replicas to the same rack in the same zone, and so on.
```

Author: It's not recommended to use "and so on" or "etc" according to[ the google style guide](https://developers.google.com/style/word-list#letter-e) @TomShawn 
> Avoid both etc. and and so on wherever possible, but if you really need to use one, use etc. Always include the period, even if a comma follows immediately after.

Author: Which one shall we choose? "data zone" or "data center"
see @disksing 's [comment](https://github.com/pingcap/docs/pull/3825#discussion_r484268468)

Reviewer: @ireneontheway Comment changed, PTAL.

Reviewer 2: I think data center is more widely used.

Author: ```suggestion
Similarly, when `isolation-level` is set to `rack`, the minimum isolation level is different racks in the same data center. Under this setting, if isolation can be guaranteed at the zone layer, it will be guaranteed first. Only when the zone level isolation can not be completed, will the scheduling in the same zone and the same rack be considered to avoid.
```
Is it ok to change "level" into "layer"?

> In summary, PD maximizes the disaster recovery of the cluster according to the current topology. Therefore, if you want to achieve a certain level of disaster recovery, deploy more machines on different sites according to the topology than the number of `max-replicas`. Also, TiDB provides mandatory isolation level settings such as `isolation-level` to control the topological isolation level according to the scenario more flexibly.

Reviewer: ```suggestion
In summary, PD maximizes the disaster recovery of the cluster according to the current topology. Therefore, if you want to achieve a certain level of disaster recovery, deploy more machines on different sites according to the topology than the number of `max-replicas`. TiDB also provides a mandatory configuration item `isolation-level` to control the topological isolation level of data according to different cases.
```

Author: ```suggestion
In summary, PD maximizes the disaster recovery of the cluster according to the current topology. Therefore, if you want to achieve a certain level of disaster recovery, deploy more machines on different sites according to the topology than the number of `max-replicas`. TiDB also provides mandatory configuration items such as `isolation-level` to control the topological isolation level of data according to different cases.
```
I've updated my version according to the source text:
> 同时 TiDB 也提供了诸如 isolation-level 这样的强制隔离级别设置，以便更灵活地根据场景来控制对数据的拓扑隔离级别。

Reviewer: To me, `mandatory isolation level settings ` is not clear enough. It's about language clarity. For other parts, I've updated my comment, PTAL.

### Configure `location-labels` for PD

According to the description above, the label can be any key-value pair used to describe TiKV attributes. But PD cannot identify the location-related labels and the layer relationship of these labels. Therefore, you need to make the following configuration for PD to understand the TiKV node topology.

+ If the PD cluster is not started, configure `location-labels` in the PD configuration file:

    {{< copyable "" >}}

    ```toml
    [replication]
    location-labels = ["zone", "rack", "host"]
    ```

+ If the PD cluster is already initialized, use the pd-ctl tool to make online changes:

    {{< copyable "shell-regular" >}}

    ```bash
    pd-ctl config set location-labels zone,rack,host
    ```

The `location-labels` configuration is an array of strings, and each item corresponds to the key of TiKV `labels`. The sequence of each key represents the layer relationship of different labels.

> **Note:**
>
> You must configure `location-labels` for PD and `labels` for TiKV at the same time for the configurations to take effect. Otherwise, PD does not perform scheduling according to the topology.

### Configure `isolation-level` for PD

> If `location-labels` has been configured, you can further enhance the topological isolation requirements on TiKV clusters by configuring `isolation-level` in the PD configuration file:

Reviewer: ```suggestion
If `location-labels` has been configured, you can further enhance the topological isolation requirements on TiKV clusters by configuring `isolation-level` in the PD configuration file.
```

> Assume that you have made a three-layer cluster topology by configuring `location-labels` according to the instructions above: zone -> rack -> host, and have configured the `isolation-level` as follows:

Reviewer: ```suggestion
Assume that you have made a three-layer cluster topology by configuring `location-labels` according to the instructions above: zone -> rack -> host, you can configure the `isolation-level` to `zone` as follows:
```

{{< copyable "" >}}

```toml
[replication]
isolation-level = "zone"
```

If the PD cluster is already initialized, you need to use the pd-ctl tool to make online changes:

{{< copyable "shell-regular" >}}

```bash
pd-ctl config set isolation-level zone
```

The `location-level` configuration is an array of strings, which needs to correspond to a key of `location-labels`. This parameter limits the minimum and mandatory isolation level requirements on TiKV topology clusters.

> **Note:**
>
> `isolation-level` is empty by default, which means there is no mandatory restriction on the isolation level. To set it, you need to configure `location-labels` for PD and ensure that the value of `isolation-level` is one of `location-labels` names.

> 

Reviewer: @ireneontheway Please align https://github.com/pingcap/docs-cn/pull/4502 in this document, thanks!

Author: Done in https://github.com/pingcap/docs/pull/3825/commits/dbd532133b09dbae13883a698ef08a13a73906ff

---
title: Schedule Replicas by Topology Labels

> aliases: ['/docs/dev/how-to/deploy/geographic-redundancy/location-awareness/','/docs/dev/location-awareness/']

Reviewer: ```suggestion
aliases: ['/docs/dev/location-awareness/','/docs/dev/how-to/deploy/geographic-redundancy/location-awareness/','/tidb/dev/location-awareness']
```

Author: Done in https://github.com/pingcap/docs/pull/3825/commits/18a44741a141609bd277bb8d3cdfec57199f8a2d

---

# Schedule Replicas by Topology Labels

To improve the high availability and disaster tolerance of TiDB clusters, it is recommended that TiKV nodes are physically distributed as much as possible. For example, TiKV nodes can be distributed on different racks or even in different data centers. According to the topology information of TiKV, the PD scheduler automatically performs scheduling at the background to isolate the replicas of Regions as much as possible, thereby maximizing the capability for disaster recovery.

> To make this mechanism effective, you need to report the topology information of the cluster, especially the TiKV location, to PD for proper configuration during deployment. Before you begin, see [Deploy TiDB Using TiDB Ansible (Recommended)](/online-deployment-using-ansible.md) and [Deploy TiDB Using Docker](/test-deployment-using-docker.md).

Reviewer: Outdated information
```suggestion
To make this mechanism effective, you need to report the topology information of the cluster, especially the TiKV location, to PD for proper configuration during deployment. Before you begin, see [Deploy TiDB Using TiUP](/tidb/dev/production-deployment-using-tiup.md).
```

## Configure `labels` based on the cluster topology

> ### Set the `label` configuration of TiKV

Reviewer: Outdated information
```suggestion
### Set the `labels` configuration of TiKV
```

> **Note:**
>
> You must configure `location-labels` for PD and `labels` for TiKV at the same time for `labels` to take effect. Otherwise, PD does not schedule according to the topology.

### Use TiDB Ansible to deploy a cluster

You can directly set location related configuration in the `inventory.ini` file when using TiDB Ansible to deploy a cluster. `tidb-ansible` generates the corresponding TiKV and PD configuration files during deployment.

The following example defines a double-level topology of `zone/host`. The TiKV of the cluster is distributed in three zones, with two hosts in each zone. z1 is deployed with two TiKV instances per host, and z2 and z3 are deployed with 1 instance per host.

```
[tikv_servers]
# z1
tikv-1 labels="zone=z1,host=h1"
tikv-2 labels="zone=z1,host=h1"
tikv-3 labels="zone=z1,host=h2"
tikv-4 labels="zone=z1,host=h2"
# z2
tikv-5 labels="zone=z2,host=h1"
tikv-6 labels="zone=z2,host=h2"
# z3
tikv-7 labels="zone=z3,host=h1"
tikv-8 labels="zone=z3,host=h2"
[pd_servers:vars]

> location_labels = ["zone", "rack"]

Reviewer: @ireneontheway There is quite a bit of outdated information in your translated `schedule-replicas-by-topology-labels.md` file. The PR https://github.com/pingcap/docs-cn/pull/3174 you translate was merged 3 months ago and changes are made to this file during the 3 months. Would you check [the change history](https://github.com/pingcap/docs-cn/commits/master/schedule-replicas-by-topology-labels.md) and update your PR? Please make sure you are aligning the latest content instead of the content composed 3 months ago? **Please base your translation on the latest version (upstream).**
```suggestion
location_labels = ["zone", "host"]
```

Author: Mainly done in https://github.com/pingcap/docs/pull/3825/commits/eb1120e75f1b2edbb57f289125cbebea1cee0dc7
